From 7310f485c5dbd51cbacc6415ed75f0c949bf8a73 Mon Sep 17 00:00:00 2001
From: Aleksandrs Vinarskis <alex.vinarskis@gmail.com>
Date: Thu, 3 Apr 2025 01:12:50 +0200
Subject: [PATCH] WIP: arm64: dts: qcom: HDMI support on Asus Zenbook A14

HDMI connector is wired via Parade PS185HDM bridge IC to
``usb_1_ss2_qmpphy`. Enable qmphy and desribe the connector.

WIP:
* HPD event works
* i2c bus gets populated, EDID can be manually read
* Cold boot works _most_ of the time
* Hot plug rarely works, as EDID info is not read out correctly by DP
  driver. Same happens when attempting to change resolution on working
  video output following the coldboot - device dissapears. My
  hypothesis is that HPD is received too fast, before AUX is ready for
  EDID readout

Not-Signed-off-by: Aleksandrs Vinarskis <alex.vinarskis@gmail.com>
---
 .../boot/dts/qcom/x1-asus-zenbook-a14.dtsi    | 58 +++++++++++++++++++
 1 file changed, 58 insertions(+)

diff --git a/arch/arm64/boot/dts/qcom/x1-asus-zenbook-a14.dtsi b/arch/arm64/boot/dts/qcom/x1-asus-zenbook-a14.dtsi
index ba32e2d5a727..195c2563189a 100644
--- a/arch/arm64/boot/dts/qcom/x1-asus-zenbook-a14.dtsi
+++ b/arch/arm64/boot/dts/qcom/x1-asus-zenbook-a14.dtsi
@@ -48,6 +48,23 @@ wcd938x: audio-codec {
 		#sound-dai-cells = <1>;
 	};
 
+	dp-connector {
+		compatible = "dp-connector";
+		label = "hdmi";
+		type = "full-size";
+
+		hpd-gpios = <&tlmm 126 GPIO_ACTIVE_HIGH>;
+
+		pinctrl-0 = <&hdmi_hpd_default>;
+		pinctrl-names = "default";
+
+		port {
+			dp_connector_in: endpoint {
+				remote-endpoint = <&usb_1_ss2_qmpphy_out>;
+			};
+		};
+	};
+
 	gpio-keys {
 		compatible = "gpio-keys";
 
@@ -1002,6 +1019,15 @@ &mdss_dp1_out {
 	link-frequencies = /bits/ 64 <1620000000 2700000000 5400000000 8100000000>;
 };
 
+&mdss_dp2 {
+	status = "okay";
+};
+
+&mdss_dp2_out {
+	data-lanes = <0 1>;
+	link-frequencies = /bits/ 64 <1620000000 2700000000 5400000000 8100000000>;
+};
+
 &mdss_dp3 {
 	compatible = "qcom,x1e80100-dp";
 	/delete-property/ #sound-dai-cells;
@@ -1273,6 +1299,12 @@ hdtl_default: hdtl-default-state {
 		function = "gpio";
 	};
 
+	hdmi_hpd_default: hdmi-hpd-state {
+		pins = "gpio126";
+		function = "gpio";
+		bias-pull-down;
+	};
+
 	kybd_default: kybd-default-state {
 		pins = "gpio67";
 		function = "gpio";
@@ -1466,6 +1498,32 @@ &usb_1_ss1_qmpphy_out {
 	remote-endpoint = <&retimer_ss1_ss_in>;
 };
 
+&usb_1_ss2 {
+	status = "okay";
+};
+
+&usb_1_ss2_dwc3 {
+	dr_mode = "host";
+};
+
+&usb_1_ss2_hsphy {
+	vdd-supply = <&vreg_l3j_0p8>;
+	vdda12-supply = <&vreg_l2j_1p2>;
+
+	status = "okay";
+};
+
+&usb_1_ss2_qmpphy {
+	vdda-phy-supply = <&vreg_l2j_1p2>;
+	vdda-pll-supply = <&vreg_l2d_0p9>;
+
+	status = "okay";
+};
+
+&usb_1_ss2_qmpphy_out {
+	remote-endpoint = <&dp_connector_in>;
+};
+
 &usb_mp {
 	status = "okay";
 };
-- 
2.45.2

